README.md
# NeuronTracking

* Get all the image data into a 3-D (X,Y,I) Float32 array.
* include("zmodel.jl")
* ztrack(data, first, last)
* Make sure "first" shows the neurons clearly for best results.


ChangeLog
2016-05-30  Deniz Yuret  <dyuret@ku.edu.tr>

	* shift: Vahdet says there is a slight xy shift every z-level,
	which causes the 17-18 jump.  How much is the shift?
	julia> ztrack(img32,18,57)
	(37,[316,235,16,31,237,340,15,29,320,286,20,37])
	(38,[269,215,17,37,221,338,20,28,287,275,15,29])
	diff:  47  20  -1  -6  16  2  -5  1  33  11  5  8

	Looking at different transitions, I can't get anything consistent.
	With visual inspection I got dx=2.5, dy=1.0 pixels per frame drift.

2016-05-29  Deniz Yuret  <dyuret@ku.edu.tr>

	* gauss.jl: Tried exponential and power law, compared to gaussian.
	Exponential is worse, power law is better.  The z/x ratio is
	consistently around 1.57, i.e. one z level is equivalent to 1.57 x
	pixels.  For Gaussian, one stdev on xy plane is 11.3 pixels, in z
	direction 7.3 levels.  For zmodel, this corresponds to a Lambda of
	2053 and ZX=1.5.  For 1/(1+(L/2)r^2) power-law model, L is 8160.
	Or if we use 1/(1+(r/R)^2), then R is 8 pixels.  Compare with
	sigma=11.3 pixels with the Gaussian.  Square law drops to half the
	peak value in 8 pixels, Gaussian in 13.4 pixels.

	* zmodel.jl: fixed drawing.  still suspect hyperparameters.  need
	to fit gaussian in 3D to see the exact centers and the z/x
	relation.

2016-05-28  Deniz Yuret  <dyuret@ku.edu.tr>

	* zmodel.jl: first successful tracking.  need to play with
	hyperparameters to optimize.  loses centers between blocks of 20.
	need to draw circles instead of adding two images to see what is
	going on.  need to make use of relative location info, currently
	each neuron is allowed to move independently in the model.

	* z-model: Looking at the brightness of the same neuron across
	levels, the optimal lambda_z~3500.  Given that lambda_x = 2000,
	this corresponds to equivalent dz=0.756dx.  i.e. roughly the same
	decrease is observed when we are 4 pixels away in x direction or 3
	pixels (levels) in the z direction.  I tried it again with raw
	pixel values instead of fitted activations and got 5700.  Better
	make this a parameter.  Fit a whole 20 image block [18:37] to
	optimize the parameters?  Or just write the sequential stuff.  But
	how can we initialize z values?

	* view: When the image on the canvas changes, the pixel values of
	the original image are still reported.  Debug
	display.jl:updatexylabel.

	Can use the transpose option for canvas.

	* jumps: The scenes jump at 18,38,58 etc, which is inconsistent
	with the z-levels from askin.



drift.jl
include("zmodel.jl")
_dx = 2.5
_dy = 1.0
_d = Array(Float32,512,512)

dlimit(a::Int,b::Int,c::Int)=(a<b ? b : a>c ? c : a)

function dview{T}(data::Array{T,3}, img::Int; temp::Matrix{T}=_d, dx=_dx, dy=_dy)
    z = mod1(img-Z0,Z)
    dx = round(Int64,dx * z)
    dy = round(Int64,dy * z)
    X,Y = size(temp)
    @inbounds for x=1:X, y=1:Y
        xx = dlimit(x+dx, 1, X)
        yy = dlimit(y+dy, 1, Y)
        temp[x,y] = data[xx,yy,img]
    end
    zview(temp)
end

function dfix{T}(data::Array{T,3}; temp::Matrix{T}=_d, dx=_dx, dy=_dy, out::Array{T,3}=similar(data))
    @assert dx>=0 && dy>=0
    copy!(out,data)
    X,Y,I = size(data)
    for i=1:I
        z = mod1(i-Z0,Z)
        x = round(Int64,dx * z)
        y = round(Int64,dy * z)
        copy!(sub(out,1:X-x,1:Y-y,i),sub(data,x+1:X,y+1:Y,i))
    end
    return out
end


gauss.jl
# Load _g1 from neuron1.jld which is a 3D z-aligned box holding one neuron

function gloss1(par)            # gaussian, fixed a0
    # Minimum: 2.544195 at [1.0279446 1.04854 0.47514126 1.6086882 1.5215918 3.7772746 0.06005877]
    global _g1,_g2
    isdefined(:_g1) || error("Define target _g1")
    isdefined(:_g2) || (_g2 = similar(_g1))
    a0 = 0.03
    (xi,yi,zi,rx,ry,rz,ai)=par
    for x=1:41,y=1:41,z=1:20
        _g2[x,y,z] = a0 + ai * exp(-rx*(xi-x/20)^2 - ry*(yi-y/20)^2 - rz*(zi-z/20)^2)
    end
    vecnorm(broadcast!(-, _g2, _g2, _g1))
end

function gloss2(par)            # gaussian a0 optimized
    # Minimum: 2.318399 at [1.0074097f0,1.0313389f0,0.4666347f0,4.5696244f0,4.1360598f0,11.593927f0,0.07910681f0,0.04300439f0]
    global _g1,_g2
    isdefined(:_g1) || error("Define target _g1")
    isdefined(:_g2) || (_g2 = similar(_g1))
    (xi,yi,zi,rx,ry,rz,ai,a0)=par
    for x=1:41,y=1:41,z=1:20
        _g2[x,y,z] = a0 + ai * exp(-rx*(xi-x/20)^2 - ry*(yi-y/20)^2 - rz*(zi-z/20)^2)
    end
    vecnorm(broadcast!(-, _g2, _g2, _g1))
end

function gloss3(par)            # exponential fixed a0
    # Minimum: 2.554641 at [1.0148009f0,1.0381743f0,0.46588328f0,1.3417006f0,1.2868823f0,1.6828519f0,0.09428701f0]
    global _g1,_g2
    isdefined(:_g1) || error("Define target _g1")
    isdefined(:_g2) || (_g2 = similar(_g1))
    (xi,yi,zi,rx,ry,rz,ai)=par
    a0 = 0.03
    for x=1:41,y=1:41,z=1:20
        _g2[x,y,z] = a0 + ai * exp(-rx*abs(xi-x/20) - ry*abs(yi-y/20) - rz*abs(zi-z/20))
    end
    vecnorm(broadcast!(-, _g2, _g2, _g1))
end

function gloss4(par)            # exponential a0 optimized
    # Minimum: 2.455714 at [1.0124047f0,1.0313879f0,0.4658681f0,2.5158482f0,2.3875716f0,3.5193841f0,0.1388294f0,0.041223526f0]
    global _g1,_g2
    isdefined(:_g1) || error("Define target _g1")
    isdefined(:_g2) || (_g2 = similar(_g1))
    (xi,yi,zi,rx,ry,rz,ai,a0)=par
    for x=1:41,y=1:41,z=1:20
        _g2[x,y,z] = a0 + ai * exp(-rx*abs(xi-x/20) - ry*abs(yi-y/20) - rz*abs(zi-z/20))
    end
    vecnorm(broadcast!(-, _g2, _g2, _g1))
end

function gloss5(par)            # exp(-r^p) with fixed a0
    # Minimum: 2.525750 at [1.020521f0,1.0451639f0,0.46920383f0,1.4722148f0,1.402376f0,2.522549f0,0.07197181f0,1.4944962f0]
    global _g1,_g2
    isdefined(:_g1) || error("Define target _g1")
    isdefined(:_g2) || (_g2 = similar(_g1))
    (xi,yi,zi,rx,ry,rz,ai,p)=par
    a0 = 0.03
    for x=1:41,y=1:41,z=1:20
        _g2[x,y,z] = a0 + ai * exp(-rx*abs(xi-x/20)^p - ry*abs(yi-y/20)^p - rz*abs(zi-z/20)^p)
    end
    vecnorm(broadcast!(-, _g2, _g2, _g1))
end

function gloss6(par)            # exp(-r^p) with a0 optimized
    # Minimum: 2.309404 at [1.0071558f0,1.0304579f0,0.46796566f0,6.200063f0,5.4875193f0,20.799503f0,0.06861037f0,0.04346131f0,2.4839554f0]
    global _g1,_g2
    isdefined(:_g1) || error("Define target _g1")
    isdefined(:_g2) || (_g2 = similar(_g1))
    (xi,yi,zi,rx,ry,rz,ai,a0,p)=par
    for x=1:41,y=1:41,z=1:20
        _g2[x,y,z] = a0 + ai * exp(-rx*abs(xi-x/20)^p - ry*abs(yi-y/20)^p - rz*abs(zi-z/20)^p)
    end
    vecnorm(broadcast!(-, _g2, _g2, _g1))
end

function gloss7(par)            # 1/(1+r^p) with fixed a0
    # Minimum: 2.351727 at [1.0172533f0,1.0372907f0,0.46628773f0,8.777429f0,8.041203f0,20.366737f0,0.11055235f0,1.799353f0]
    global _g1,_g2
    isdefined(:_g1) || error("Define target _g1")
    isdefined(:_g2) || (_g2 = similar(_g1))
    (xi,yi,zi,rx,ry,rz,ai,p)=abs(par)
    a0 = 0.03
    for x=1:41,y=1:41,z=1:20
        _g2[x,y,z] = a0 + ai / (1 + (rx*(xi-x/20)^2 + ry*(yi-y/20)^2 + rz*(zi-z/20)^2)^(p/2))
    end
    vecnorm(broadcast!(-, _g2, _g2, _g1))
end

function gloss8(par)            # 1/(1+r^p) with optimized a0
    # Minimum: 2.318999 at [1.0090305f0,1.0319712f0,0.4678052f0,5.4503393f0,4.9420156f0,13.582409f0,0.07206944f0,0.04085429f0,3.5884995f0]
    global _g1,_g2
    isdefined(:_g1) || error("Define target _g1")
    isdefined(:_g2) || (_g2 = similar(_g1))
    (xi,yi,zi,rx,ry,rz,ai,a0,p)=abs(par)
    for x=1:41,y=1:41,z=1:20
        _g2[x,y,z] = a0 + ai / (1 + (rx*(xi-x/20)^2 + ry*(yi-y/20)^2 + rz*(zi-z/20)^2)^(p/2))
    end
    vecnorm(broadcast!(-, _g2, _g2, _g1))
end

function gloss9(par)            # 1/(1+r^2) with fixed a0
    # Minimum: 2.355760 at [1.0174382f0,1.0378766f0,0.46776757f0,6.482576f0,5.9682374f0,15.343796f0,0.09882537f0]
    global _g1,_g2
    isdefined(:_g1) || error("Define target _g1")
    isdefined(:_g2) || (_g2 = similar(_g1))
    (xi,yi,zi,rx,ry,rz,ai)=abs(par)
    a0 = 0.03; p = 2
    for x=1:41,y=1:41,z=1:20
        _g2[x,y,z] = a0 + ai / (1 + (rx*(xi-x/20)^2 + ry*(yi-y/20)^2 + rz*(zi-z/20)^2)^(p/2))
    end
    vecnorm(broadcast!(-, _g2, _g2, _g1))
end

function gloss10(par)            # 1/(1+r) with fixed a0
    # Minimum: 2.457183 at [1.0198932f0,1.0372633f0,0.46870422f0,559.0978f0,656.0105f0,1288.5681f0,0.42343268f0]
    global _g1,_g2
    isdefined(:_g1) || error("Define target _g1")
    isdefined(:_g2) || (_g2 = similar(_g1))
    (xi,yi,zi,rx,ry,rz,ai)=abs(par)
    a0 = 0.03; p = 1
    for x=1:41,y=1:41,z=1:20
        _g2[x,y,z] = a0 + ai / (1 + (rx*(xi-x/20)^2 + ry*(yi-y/20)^2 + rz*(zi-z/20)^2)^(p/2))
    end
    vecnorm(broadcast!(-, _g2, _g2, _g1))
end

# optimize(gloss, Float32[1.0,1.0,0.5,1.0,1.0,1.0,.1]; show_trace=true, show_every=10, method=LBFGS())


model.jl
# The first attempt did not work.  Try ignoring the background pixels for loss or use cutoff.
# Centers at the edge have a large gradient.
# Small centers have no gradient.
# Few bright pixels close by beat lots of bright pixels far away and capture the center.
# We need to initialize well and use the model to correct for small shifts.

# Forward model:
# Input: (3,k) matrix of x,y,a triples.
# Output: (n,n) matrix of brightness values.
# Using picture coordinates.  [0,1]->[0,512]
# Ignore 3D for now.
# x[1,k],x[2,k] in [0,1], x[3,k] in [.2,.4]?

lambda = 2000.0
baseline = 0.03
coor1 = Array(Float32,512,512); for i=1:512,j=1:512; coor1[i,j]=i/512; end
coor2 = Array(Float32,512,512); for i=1:512,j=1:512; coor2[i,j]=j/512; end
temp = [ Array(Float32,512,512) for i=1:10 ]

zfactor = 2
zlevel(z)=mod1(z+3,20)

function mforw{T}(par::Matrix{T},img::Matrix{T})
    fill!(img, baseline)
    for k=1:size(par,2)
        (x,y,a) = par[:,k]
        broadcast!(-, temp[1], coor1, x)
        broadcast!(*, temp[1], temp[1], temp[1])
        broadcast!(-, temp[2], coor2, y)
        broadcast!(*, temp[2], temp[2], temp[2])
        broadcast!(+, temp[2], temp[1], temp[2])
        broadcast!(*, temp[2], temp[2], -lambda)
        broadcast!(exp, temp[2], temp[2])
        broadcast!(*, temp[2], temp[2], a)
        broadcast!(+, img, img, temp[2])
    end
    return img
end

function mforw0{T}(par::Matrix{T},img::Matrix{T})
    _,K = size(par)
    I,J = size(img)
    fill!(img, baseline)
    @inbounds @simd for j=1:J
        @inbounds for i=1:I
            @inbounds for k=1:K
                (x,y,a)=par[:,k]
                img[i,j] += a * exp(-lambda * ((x-i/I)^2 + (y-j/J)^2))
            end
        end
    end
    return img
end

# Backward gradient:
function mback{T}(par::Matrix{T},pred::Matrix{T},gold::Matrix{T},diff::Matrix{T})
    fill!(diff,0)
    broadcast!(-, temp[10], pred, gold)
    for k=1:size(par,2)
        (x,y,a) = par[:,k]
        broadcast!(-, temp[1], coor1, x)
        broadcast!(-, temp[2], coor2, y)
        broadcast!(*, temp[3], temp[1], temp[1])
        broadcast!(*, temp[4], temp[2], temp[2])
        broadcast!(+, temp[5], temp[3], temp[4])
        broadcast!(*, temp[5], temp[5], -lambda)
        broadcast!(exp, temp[5], temp[5])
        broadcast!(*, temp[5], temp[5], temp[10]) # delta*exp(-lambda*d2)
        broadcast!(*, temp[6], temp[5], 2*a*lambda)
        broadcast!(*, temp[1], temp[1], temp[6])
        broadcast!(*, temp[2], temp[2], temp[6])
        diff[1,k] += sum(temp[1])
        diff[2,k] += sum(temp[2])
        diff[3,k] += sum(temp[5])
    end
    return diff
end

function mback0{T}(par::Matrix{T},pred::Matrix{T},gold::Matrix{T},diff::Matrix{T})
    fill!(diff,0)
    _,K = size(par)
    I,J = size(pred)
    for j=1:J
        for i=1:I
            delta = pred[i,j]-gold[i,j]
            for k=1:K
                (x,y,a) = par[:,k]
                di = (i/I - x)
                dj = (j/J - y)
                d2 = di*di + dj*dj
                tmp1 = delta * exp(-lambda * d2)
                tmp2 = 2 * a * lambda * tmp1
                diff[1,k] += tmp2 * di
                diff[2,k] += tmp2 * dj
                diff[3,k] += tmp1
            end
        end
    end
    diff
end

# Loss function:

function mloss{T}(ypred::Matrix{T},ygold::Matrix{T})
    0.5*vecnorm(ypred-ygold)^2
end

function mlossxy{T}(x::Matrix{T},ygold::Matrix{T})
    ypred = similar(ygold)
    mforw(x, ypred)
    mloss(ypred, ygold)
end

# Initialize randomly:
# At the edges it shoots out to match dark pixels.
# function minit{T}(x::Matrix{T}=Array(Float32,2,3))
#     for k=1:length(x)
#         x[k] = 0.2*(2*rand()-1)+0.6
#     end
#     x
# end

function minit{T}(y::Matrix{T})
    x = Array(T,3,3)
    c = centers(y)
    for k=1:size(x,2)
        x[1,k] = c[k][1]/size(y,1)
        x[2,k] = c[k][2]/size(y,2)
        x[3,k] = c[k][3]-baseline
    end
    x
end

using Base.LinAlg.axpy!
using ImageView

function mview(y;cmax=maximum(y))
    global _mview
    try
        view(_mview,y')
    catch e
        (_mview,_) = view(y',clim=[0,cmax],name="mview")
    end
end

# mview(y;cmax=alpha)=view(y',clim=[0,cmax])
# mview(c,y;cmax=alpha)=view(c,y',clim=[0,cmax])

# Training:
function mtrain{T}(x::Matrix{T}, ygold::Matrix{T}; iter=50, gmin=1.0, lr=0.0004)
    global yview = copy(ygold)
    alpha = maximum(ygold)/2
    mview(yview)
    ypred = similar(ygold)
    dx = similar(x)
    for i=1:iter
        mforw(x, ypred)
        loss = Float32(mloss(ypred, ygold))
        mback(x, ypred, ygold, dx)
        axpy!(-lr, dx, x)
        nx = vecnorm(dx)
        println((i,loss,nx,round(Int64,512*x[:])))
        axpy!(1, ypred, copy!(yview,ygold))
        mview(yview)
        nx < gmin && break
    end
    x
end

# Gradient check:
function mcheck{T}(x::Matrix{T}, ygold::Matrix{T}; eps=1e-3)
    ypred = similar(ygold)
    dx = similar(x)
    mforw(x, ypred)
    mback(x, ypred, ygold, dx)
    for i=1:length(x)
        xi = x[i]
        x[i] = xi - eps
        mforw(x, ypred)
        f1 = mloss(ypred, ygold)
        x[i] = xi + eps
        mforw(x, ypred)
        f2 = mloss(ypred, ygold)
        x[i] = xi
        df = (f2-f1)/(2*eps)
        println((i,df,dx[i]))
    end
end



explore.jl
using ImageMagick,FileIO,JLD,FixedPointNumbers
const imgdir = "/mnt/ai/work/askin/150723_exp3_GCAMP_Images"

# Save all images as a list of images
function img2jld()
    global img = Any[]
    for i=1:7002
        i%100==0 && print(".")
        push!(img, load("$imgdir/$i.tiff"))
    end
    println(); info("Saving...")
    JLD.save("img.jld","img",img)
end

# Save all images as a 512x512x7002 array of UFixed16 values
function img3jld(img)
    global img3=Array(UFixed16,512,512,7002)
    for i=1:7002; copy!(img3,1+(i-1)*512*512,img[i].data,1,512*512); end
    JLD.save("img3.jld","img3",img3)
end

# Find centers as local maxima of pixel values: doesn't work very well
function centers{T}(x::Matrix{T}; r=30)
    c = Any[]
    for i=1:size(x,1)
        for j=1:size(x,2)
            localmax(x,i,j,r) && push!(c, (i,j,x[i,j]))
        end
    end
    return sort(c, by=(x->x[3]), rev=true)
end

function findmax2(a)
    v,i = findmax(a)
    v,ind2sub(a,i)
end

function centers32{T}(x::Matrix{T}; r=32)
    c = Any[]
    for j=1:r:size(x,2)
        jj=min(j+r-1,size(x,2))
        for i=1:r:size(x,1)
            ii=min(i+r-1,size(x,1))
            push!(c, ((i+ii)/2,(j+jj)/2,mean(sub(x,i:ii,j:jj))))
        end
    end
    return sort(c, by=(x->x[3]), rev=true)
end

function localmax{T}(x::Matrix{T}, i, j, r)
    for ii=max(1,i-r):min(size(x,1),i+r)
        for jj=max(1,j-r):min(size(x,2),j+r)
            if x[i,j] < x[ii,jj]
                return false
            end
        end
    end
    return true
end

# Find the image with the max pixel for every 20 image block
function findmax20()
    zz = Int[]
    for i=0:20:size(img3,3)
        d = img3[:,:,i+1:min(i+20,7002)]
        (v,n) = findmax(d)
        (x,y,z) = ind2sub(d,n)
        push!(zz,z)
    end
    zz
end

# Find variation in activation looking at the value of the max pixel for every 20 image block
function findmaxpix()
    zz = Float32[]
    for i=0:20:size(img3,3)
        d = img3[:,:,i+1:min(i+20,7002)]
        push!(zz,maximum(d))
    end
    zz
end

# Find variation based on distance around the max pixel in every block
function pixeldistance(; r=50)
    dr = Float32[]; dv = Float32[]
    for i=0:20:size(img3,3)
        d = img3[:,:,i+1:min(i+20,7002)]
        (v,n) = findmax(d)
        (v > 0.5) && continue
        (x,y,z) = ind2sub(d,n)
        for xx = max(1,x-r):min(512,x+r)
            for yy = max(1,y-r):min(512,y+r)
                vv = d[xx,yy,z]
                push!(dr, sqrt((x-xx)^2 + (y-yy)^2))
                push!(dv, Float32(vv)-Float32(v))
            end
        end
    end
    (dr, dv)
end

function brighten(img)
    c = copy(img)
    for i=1:length(c.data)
        c.data[i] = min(1, 5*c.data[i])
        # c.data[i] = max(0, 1-4*c.data[i])
    end
    return c
end


zmodel.jl
# model parameters
A0 = 0.03
Lambda = 2000f0                 # brightness falls with exp(-0.5*Lambda*r^2)
# empirical lambda_x = 1f-6, lambda_a = 1f-2
Tau = Float32[0,0,1f-2,1f-4]  # how much x,y,z,a change between timesteps

N = 3                           # number of neurons
X = 512                         # image width, assuming square X,X images
Z = 20                          # number of layers
ZX = 1.5                        # one z layer is ZX pixels in distance
Z0 = 17                         # Z0+1:Z0+Z is one block
Z1 = Z0+div(Z,2)                # middle of the block
X1drift = 2.5                   # image drift in pixels for first dim
X2drift = 1.0                   # image drift in pixels for second dim

# we map 1:X to 0:1 in the xy directions.
X1 = Array(Float32,X,X); for i=1:X,j=1:X; X1[i,j]=i/X; end
X2 = Array(Float32,X,X); for i=1:X,j=1:X; X2[i,j]=j/X; end

# we map 1:Z to 0:ZX*Z/X in the z direction. some conversions:
# p: pixels, i: images, z: z levels, x: x units
p2x(p::Int)=Float32(p/X)
x2p{T}(x::T)=round(Int,x*X)
z2x(z::Int)=Float32(z*ZX/X)
x2z{T}(x::T)=round(Int,x*X/ZX)
i2z(i::Int)=mod1(i-Z0,Z)
i2x(i::Int)=z2x(i2z(i))

# alloc some temporary arrays
M = [ Array(Float32,X,X) for i=1:20 ]
W = [ Array(Float32,4,N) for i=1:10 ]


function zforw{T}(par::Matrix{T},out::Matrix{T}=M[1]; img=Z1,
                  m1::Matrix{T}=M[2], m2::Matrix{T}=M[3])
    fill!(out, A0)                              # A0 is the background brightness
    X3 = i2x(img)                             # X1,X2,X3 represent pixel positions
    for k=1:size(par,2)
        (x1,x2,x3,a) = par[:,k]                 # x1,x2,x3 represent center position, a is activation
        broadcast!(-, m1, X1, x1)
        broadcast!(*, m1, m1, m1)               # (X1-x1)^2
        broadcast!(-, m2, X2, x2)
        broadcast!(*, m2, m2, m2)               # (X2-x2)^2
        broadcast!(+, m2, m1, m2)               # +(X1-x1)^2
        broadcast!(+, m2, m2, (x3-X3)^2) 	# +(X3-x3)^2
        broadcast!(*, m2, m2, -Lambda/2)
        broadcast!(exp, m2, m2)                 # exp(-Lambda/2 * d2)
        broadcast!(*, m2, m2, a)                # ai * exp(-Lambda/2 * d2)
        broadcast!(+, out, out, m2)             # a0 + sum ai * exp(-Lambda/2 * d2)
    end
    return out
end

function zloss{T}(curr::Matrix{T},prev::Matrix{T},pred::Matrix{T},gold::Matrix{T}; img=Z1,
                  xdiff::Matrix{T}=W[1], ydiff::Matrix{T}=M[10])
    broadcast!(-, xdiff, curr, prev)
    broadcast!(*, xdiff, xdiff, xdiff)
    broadcast!(*, xdiff, xdiff, Tau)
    broadcast!(-, ydiff, pred, gold)
    broadcast!(*, ydiff, ydiff, ydiff)
    return (mean(ydiff) + sum(xdiff))/2
end

# Backward gradient:
function zback{T}(curr::Matrix{T},prev::Matrix{T},pred::Matrix{T},gold::Matrix{T},diff::Matrix{T}=M[4]; img=Z1,
                  delta::Matrix{T}=M[5], dx1::Matrix{T}=M[6], dx2::Matrix{T}=M[7], m1::Matrix{T}=M[8], m2::Matrix{T}=M[9])
    X3 = i2x(img)
    broadcast!(-, diff, curr, prev)
    broadcast!(*, diff, diff, Tau)
    broadcast!(-, delta, pred, gold)
    for k=1:size(curr,2)
        (x1,x2,x3,a) = curr[:,k]
        broadcast!(-, dx1, X1, x1)
        broadcast!(-, dx2, X2, x2)
        dx3 = X3 - x3
        broadcast!(*, m1, dx1, dx1)
        broadcast!(*, m2, dx2, dx2)
        broadcast!(+, m2, m2, m1)
        broadcast!(+, m2, m2, dx3*dx3)
        broadcast!(*, m2, m2, -Lambda/2)
        broadcast!(exp, m2, m2)
        broadcast!(*, m2, m2, delta) 
        diff[4,k] += mean(m2)                 # dJ/da = sum delta*exp(-Lambda*dist2)
        broadcast!(*, m2, m2, a*Lambda)
        diff[1,k] += mean(broadcast!(*, dx1, dx1, m2))
        diff[2,k] += mean(broadcast!(*, dx2, dx2, m2))
        diff[3,k] += mean(broadcast!(*, m2, m2, dx3))
    end
    return diff
end

# Gradient check:
function zcheck{T}(curr::Matrix{T},prev::Matrix{T},gold::Matrix{T}; img=Z1, eps=T(1e-4),
                   pred::Matrix{T}=M[11], diff::Matrix{T}=W[2])
    zforw(curr, pred, img=img)
    zback(curr, prev, pred, gold, diff, img=img)
    for i=1:length(curr)
        xi = curr[i]
        curr[i] = xi - eps
        zforw(curr, pred, img=img)
        f1 = zloss(curr, prev, pred, gold)
        curr[i] = xi + eps
        zforw(curr, pred, img=img)
        f2 = zloss(curr, prev, pred, gold)
        curr[i] = xi
        df = (f2-f1)/(2*eps)
        println((i,df,diff[i]))
    end
end

# Training:
function ztrain{T}(curr::Matrix{T}, prev::Matrix{T}, gold::Matrix{T}; img=Z1, iter=50, gmin=1f-5, gmax=1f-4, lr=0f0,
                   pred::Matrix{T}=M[12], disp::Matrix{T}=M[13], diff::Matrix{T}=W[3], prn=false)
    zdraw(curr, gold, img=img)
    for i=1:iter
        zforw(curr, pred, img=img)
        zback(curr, prev, pred, gold, diff, img=img)
        gnorm = vecnorm(diff)
        gnorm > gmax && scale!(gmax/gnorm, diff)
        lr == 0 && (@show lr = 5f-3/gnorm)
        Base.LinAlg.axpy!(-lr, diff, curr)
        zdraw(curr, gold, img=img)
        vecnorm(diff) < gmin && break
        prn && println((:img,img,:loss,zloss(curr, prev, zforw(curr,pred,img=img), gold, img=img),:gnorm,vecnorm(diff),:x,round(Int64,X*curr[:])))
    end
    loss = zloss(curr, prev, zforw(curr,pred,img=img), gold, img=img)
    println((img,i2z(img),xx2pp(curr)))
    curr
end

function xx2pp{T}(x::Matrix{T})
    a = cell(length(x))
    for i=1:4:length(x)
        a[i] = x2p(x[i])
        a[i+1] = x2p(x[i+1])
        a[i+2] = x2z(x[i+2])
        a[i+3] = round(Int,x[i+3]*0xff)
    end
    return a
end

# This loses the centers going from one block to the next.
# So we need to fix the drift using zfix.
function ztrack{T}(y::Array{T}, i1=Z1, i2=size(y,3); lr=100)
    fixed = zfix(y[:,:,i1], i1)
    prev = zinit(fixed, img=i1)
    curr = copy(prev)
    for i=i1:i2
        fixed = zfix(y[:,:,i], i)
        ztrain(curr, prev, fixed; img=i, lr=lr)
        copy!(prev, curr)
    end
end

function zfix{T}(img::Array{T,2}, i::Int; dx=X1drift, dy=X2drift, out::Array{T,2}=M[15])
    copy!(out,img)
    X,Y = size(img)
    z = i2z(i)
    x = round(Int,dx * z); y = round(Int,dy * z)
    ox1 = max(1,1-x); ox2 = min(X,X-x); oy1 = max(1,1-y); oy2 = min(Y,Y-y)
    ix1 = max(1,1+x); ix2 = min(X,X+x); iy1 = max(1,1+y); iy2 = min(Y,Y+y)
    copy!(sub(out,ox1:ox2,oy1:oy2), sub(img,ix1:ix2,iy1:iy2))
    return out
end

# Tracking
# This version starts from the middle of each block, goes back and forth
function ztrack2{T}(y::Array{T}, img1=1, img2=size(y,3);
                    curr::Matrix{T}=W[6], prev::Matrix{T}=W[7])
    while img1 < img2
        i1 = img1; while mod1(i1-Z0,Z) > 1; i1-=1; end
        i3 = img1; while mod1(i3-Z0,Z) < Z; i3+=1; end
        i2 = i1 + 10
        cent = zinit(y[:,:,i2], img=i2)
        copy!(prev, cent); copy!(curr, cent)
        for i=i2:-1:max(1,i1)
            ztrain(curr, prev, y[:,:,i], img=i)
            copy!(prev, curr)
            i==i2 && copy!(cent, curr)
        end
        copy!(prev, cent); copy!(curr, cent)
        for i=i2+1:i3
            ztrain(curr, prev, y[:,:,i], img=i)
            copy!(prev, curr)
        end
        img1 = i3+1
    end
end

function zinit{T}(y::Matrix{T};img=Z1)
    x = Array(T,4,N)
    c = zcenters(y)
    for n=1:N
        x[1,n] = c[n][1]/size(y,1)
        x[2,n] = c[n][2]/size(y,2)
        x[3,n] = i2x(img)
        x[4,n] = mean(y[c[n][1]-2:c[n][1]+2,c[n][2]-2:c[n][2]+2])-A0
    end
    return x
end

function zcenters{T}(x::Matrix{T}; r=30)
    c = Any[]
    for i=1:size(x,1)
        for j=1:size(x,2)
            zlocalmax(x,i,j,r) && push!(c, (i,j,x[i,j]))
        end
    end
    return sort(c, by=(x->x[3]), rev=true)
end

function zlocalmax{T}(x::Matrix{T}, i, j, r)
    for ii=max(1,i-r):min(size(x,1),i+r)
        for jj=max(1,j-r):min(size(x,2),j+r)
            if x[i,j] < x[ii,jj]
                return false
            end
        end
    end
    return true
end

using ImageView

function zview(y;cmax=maximum(y))
    global _zview
    try
        view(_zview,y)
    catch e
        (_zview,_) = view(y,clim=[0,cmax],name="zview",xy=["y","x"])
    end
end

using Images, Colors
using ImageView: annotate!, AnnotationText

function zdraw{T}(curr::Matrix{T}, gold::Matrix{T}; img=Z1, temp::Matrix{T}=M[14])
    global _zdraw
    copy!(temp, gold)
    X3 = i2x(img)                             # X1,X2,X3 represent pixel positions
    # f = a0 + ai * exp(-lambda/2 * d2)
    dmin = -log(0.3)/(Lambda/2)
    dmax = 1f0 # -log(0.05)/(Lambda/2)
    for k=1:size(curr,2)
        (x1,x2,x3,a) = curr[:,k]                 # x1,x2,x3 represent center position, a is activation
        X1a = max(1/X, x1-20/X); X1b = min(1, x1+20/X)
        X2a = max(1/X, x2-20/X); X2b = min(1, x2+20/X)
        for X1=X1a:1/X:X1b, X2=X2a:1/X:X2b
            d2 = (x1-X1)^2 + (x2-X2)^2 + (x3-X3)^2
            dmin <= d2 <= dmax && (temp[x2p(X1),x2p(X2)]=0.0)
        end
    end
    annot = sprint() do s
        println(s, "$img ($(i2z(img)))")
        writedlm(s, reshape(append!(Any["x","y","z","a"], xx2pp(curr)),(4,4)))
    end
    try
        a = first(values(_zdraw[1].annotations))
        a.data.string = annot
        view(_zdraw[1],temp)
    catch e
        _zdraw = view(temp, clim=[0,maximum(temp)], name="zdraw", xy=["y","x"])
        annotate!(_zdraw[1], _zdraw[2], AnnotationText(10, 10, annot, halign="left", valign="top", color=RGB(1,1,1)))
        view(_zdraw[1],temp)
    end
end
